src/main/java/de/tum/in/www1/artemis/service/GradingScaleService.java

```diff
@@ -0,0 +1,123 @@
+        if (gradingScale.getCourse() != null && gradingScale.getExam() != null) {
+            throw new BadRequestAlertException("Grading scales can't belong both to a course and an exam", "gradingScale", "gradingScaleBelongsToCourseAndExam");
+        }
+        Set<GradeStep> gradeSteps = gradingScale.getGradeSteps();
+        checkGradeStepValidity(gradeSteps);
+        for (GradeStep gradeStep : gradeSteps) {
+            gradeStep.setGradingScale(gradingScale);
+        }
+        gradingScale.setGradeSteps(gradeSteps);
+        return gradingScaleRepository.save(gradingScale);
+    }
+
+    /**
+     * Checks the validity of a grade step set and throws an exception if one of the following conditions is not fulfilled
+     * - all individuals grade steps should be in a valid format
+     * - the grade steps set should form a valid and congruent grading scale
+     *
+     * @param gradeSteps the grade steps to be checked
+     */
+    private void checkGradeStepValidity(Set<GradeStep> gradeSteps) {
+        if (gradeSteps != null) {
+            if (!gradeSteps.stream().allMatch(GradeStep::isValid)) {
+                throw new BadRequestAlertException("Not all grade steps are following the correct format.", "gradeStep", "invalidFormat");
+            }
+            else if (!gradeStepSetMapsToValidGradingScale(gradeSteps)) {
+                throw new BadRequestAlertException("Grade step set can't match to a valid grading scale.", "gradeStep", "invalidFormat");
+            }
+        }
+    }
+
+    /**
+     * Checks if the grade steps map to a valid grading scale
+     * - the grade names should all be unique for the grading scale
+     * - when ordered, all steps should fulfill valid adjacency
+     * - the first and the last element should fulfill the boundary conditions (start with 0% and end with 100%)
+     *
+     * @param gradeSteps the grade steps to be checked
+     * @return true if the grade steps map to a valid grading scale and false otherwise
+     */
+    private boolean gradeStepSetMapsToValidGradingScale(Set<GradeStep> gradeSteps) {
+        if (gradeSteps.stream().map(GradeStep::getGradeName).distinct().count() != gradeSteps.size()) {
+            return false;
+        }
+        List<GradeStep> sortedGradeSteps = gradeSteps.stream().sorted(Comparator.comparingDouble(GradeStep::getLowerBoundPercentage)).collect(Collectors.toList());
+        boolean validAdjacency = IntStream.range(0, sortedGradeSteps.size() - 1).allMatch(i -> GradeStep.checkValidAdjacency(sortedGradeSteps.get(i), sortedGradeSteps.get(i + 1)));
+        boolean validFirstElement = sortedGradeSteps.get(0).isLowerBoundInclusive() && sortedGradeSteps.get(0).getLowerBoundPercentage() == 0;
+        boolean validLastElement = sortedGradeSteps.get(sortedGradeSteps.size() - 1).isUpperBoundInclusive()
+                && sortedGradeSteps.get(sortedGradeSteps.size() - 1).getUpperBoundPercentage() == 100;
```

Comment:
Maybe you should rearrange the code for better readability, add parentheses and more line breaks. 
I'm also not sure about the first line, doesn't the `sortedGradeSteps.get(i + 1))` try to access an out of bounds element in the last step?

Reply:
It doesn't because the last index of the `range(0, sortedGradeSteps.size() - 1) is exclusive, meaning the index `i` will only count up to the second last element

Reply:
alright :+1: